### ISSUE TITLE
api methods aren't modified by middleware if they are called in a non-reactive way from prototype

### ISSUE THREAD

**Issue by LuboGocnik** [pmndrs/zustand#881]
Hello,

I am trying to use zustand with immer using the "immer middleware" from this example: [middlewareTypes.test.tsx](https://github.com/pmndrs/zustand/blob/HEAD/tests/middlewareTypes.test.tsx).

First of all, I realized that this example does not modify the **api.setState** method the same way, as it modifies the **set** method.
I am not sure if this was intentional, if yes, some comment in the example would be really great. 
Anyway, I modified also the **api.setState** and this is my code:
```typescript

type ImmerSetState<T extends State> = (partial: ((draft: Draft<T>) => void) | T, replace?: boolean) => void

export const immerMiddleware =
  <
    T extends State,
    CustomSetState extends SetState<T>,
    CustomGetState extends GetState<T>,
    CustomStoreApi extends StoreApi<T>
  >(
    config: StateCreator<T, ImmerSetState<T>, GetState<T>, CustomStoreApi>
  ): StateCreator<T, CustomSetState, CustomGetState, CustomStoreApi> =>
  (set, get, api) =>
    config(
      (partial, replace) => {
        console.log('set')
        const nextState = typeof partial === 'function' ? produce(partial as (state: Draft<T>) => T) : (partial as T)
        return set(nextState, replace)
      },
      get,
      {
        ...api,
        setState: (partial, replace) => {
          console.log('setState')
          const nextState = typeof partial === 'function' ? produce(partial as (state: Draft<T>) => T) : (partial as T)
          return api.setState(nextState, replace)
        },
      }
    )
```

As you can see, the only difference is that:

1.  I created a custom ImmerSetState type because I think that it is a little bit more readable.
2.  I copied the "immer set" function and used it for the **api.setState** method
(the only difference in that function is, that at the end I call `api.setState(nextState, replace)` instead of `set(nextState, replace)` )
3.  I added some console logs to these functions so I am 100 % sure, when do I call my modified functions and when do I call the prototype versions of these functions.

And this is the function that creates my store:
```typescript
import create, { State } from 'zustand'
import { Draft } from 'immer'
import { immerMiddleware } from './middleware'

type ImmerStore<T> = T & {
  // TS doesn't like the ImmerSetState here, so I had to remove the T option from partial
  set: (partial: (draft: Draft<T>) => void, replace?: boolean) => void
}

export const createStore = <T extends State>(args: T) =>
  create<ImmerStore<T>>(immerMiddleware((set, get, api) => ({ ...args, set, api })))
```

And this is the useStore hook:
```typescript
import { createStore } from './utils'

export interface CounterState {
  count: number
}

export interface StoreState {
  red: CounterState
  green: CounterState
  blue: CounterState
}

export const useStore = createStore<StoreState>({
  red: { count: 0 },
  green: { count: 0 },
  blue: { count: 0 },
})
```

And here is my usecase for the Red counter:
```typescript
import React, { FC, useCallback } from 'react'
import { useStore } from 'store'
import { Counter } from '../Counter'

export const RedCounter: FC = () => {
  const count = useStore((state) => state.red.count)
  const set = useStore((state) => state.set)
  const setState = useStore((state) => state.api.setState)

  const incrementSet = useCallback(
    () =>
      set((state) => {
        state.red.count += 1
      }),
    [set]
  )
  const incrementSetState = useCallback(
    () =>
      setState((state) => {
        state.red.count += 1
      }),
    [setState]
  )
  const incrementNonReactiveSetState = useCallback(
    () =>
      useStore.setState((state) => {
        state.red.count += 1
      }),
    []
  )

  return (
    // Counter is just 3 buttons + displayed count
    <Counter
      count={count}
      incrementSet={incrementSet}
      incrementSetState={incrementSetState}
      incrementNonReactiveSetState={incrementNonReactiveSetState}
    />
  )
}
```

Now when I call:

-  incrementSet: there is a "set" in a console log
- incrementSetState: there is a "setState" in a console log
- incrementNonReactiveSetState: there nothing in a console log (if I called previous functions before this one, there is an error: `index.tsx:27 Uncaught TypeError: Cannot assign to read only property 'count' of object '#<Object>'`

Which brings me to _probably_ another issue, state is protected from mutable changes only after first set call, because the produce function was used. So if someone is combining how they modify the state  (with|without immerSet) sometimes the state is protected from the mutable changes and sometimes it isn't.

**TLDR. My question is, if there is some way how to apply middleware modifications for the non-reactive methods of the useStore hook?**

**Comment by dai-shi** [comment]
Hi!

> First of all, I realized that this example does not modify the api.setState method the same way, as it modifies the set method.

I'd say it's intentional as a simple solution.
We will add `immer` middleware in v4, so if it's what people want, we can modify `api.setState` too. @devanshj What do you think?

> if there is some way how to apply middleware modifications for the non-reactive methods of the useStore hook?

You need to mutate `api` to change the api methods.
See this for example:
https://github.com/pmndrs/zustand/blob/a418fd748077c453efbff2d03641ce0af780b3c7/src/middleware/subscribeWithSelector.ts#L74-L75

**Comment by devanshj** [comment]
> We will add `immer` middleware in v4, so if it's what people want, we can modify api.setState too. @devanshj What do you think?

Yeah that's a good idea. (I think I would have already done it but it seems I missed it.)
